deepchem/models/tensorgraph/tensor_graph.py

```diff
@@ -181,29 +181,37 @@ def create_feed_dict():
       if restore:
         self.restore()
       avg_loss, n_batches = 0.0, 0.0
-      coord = tf.train.Coordinator()
       n_samples = 0
+      n_enqueued = [0]
+      final_sample = [None]
       if self.use_queue:
         enqueue_thread = threading.Thread(
             target=_enqueue_batch,
             args=(self, feed_dict_generator, self._get_tf("Graph"),
-                  self.session, coord))
+                  self.session, n_enqueued, final_sample))
         enqueue_thread.start()
       fetches = [train_op, self.loss.out_tensor]
       for feed_dict in create_feed_dict():
-        try:
-          fetched_values = self.session.run(fetches, feed_dict=feed_dict)
-          loss = fetched_values[-1]
-          avg_loss += loss
-          n_batches += 1
-          self.global_step += 1
-          n_samples += 1
-          if self.tensorboard and n_samples % self.tensorboard_log_frequency == 0:
-            summary = self.session.run(
-                self._get_tf("summary_op"), feed_dict=feed_dict)
-            self._log_tensorboard(summary)
-        except OutOfRangeError:
-          break
+        if self.use_queue:
+          # Don't let this thread get ahead of the enqueue thread, since if
+          # we try to read more batches than the total number that get queued,
+          # this thread will hang indefinitely.
+          while n_enqueued[0] <= n_samples:
+            if n_samples == final_sample[0]:
+              break
+            time.sleep(0)
+          if n_samples == final_sample[0]:
+            break
+        fetched_values = self.session.run(fetches, feed_dict=feed_dict)
+        loss = fetched_values[-1]
+        avg_loss += loss
+        n_batches += 1
+        self.global_step += 1
+        n_samples += 1
+        if self.tensorboard and n_samples % self.tensorboard_log_frequency == 0:
+          summary = self.session.run(
```

Comment:
I think we have to add a n_batches to inside this if statement.

When using a queue calculating the summary op does a pull off of the queue.

Reply:
> I think we have to add a n_batches to inside this if statement.

I'm not sure what you mean by that?

> When using a queue calculating the summary op does a pull off of the queue.

That sounds like a bug!  It means every time it logs to tensorboard, we skip training on a batch (or with the old code, two batches).  Instead of calling `run()` separately, shouldn't we just add `summary_op` to `fetches`?

Reply:
Yes it is a bug when queues are turned on -- we loose 1/self.tensorboard_log_frequency batches when training.

we should add summary_op to fetches but we don't want to eat the performance hit of computing it for every batch.  We have to add summary_op to fetches and then remove it and log it every 1/self.tensorboard_log_frequency batches.

Reply:
Ok, I'll fix that.